### ISSUE TITLE
Optional emptiness check gives incorrect error message in `map`

### ISSUE THREAD

**Issue by shas19** [uber/NullAway#307]
Currently with the OptionalEmptiness flag turned on,
this throws an error,
```
private static boolean perhaps() { return Math.random() > 0.5; }",
 void foo(Observable<Optional<String>> observable) {
              observable
                  .filter(optional -> optional.isPresent() || perhaps())
                  // BUG: Diagnostic contains: returning @Nullable expression from method with @NonNull
                  .map(optional -> optional.get());
     }
```
The error should be `Optional optional can be empty`.

**Comment by lazaroclapp** [comment]
Oh, ok, I see this one issue.

I am not sure why, maybe restrictive annotations, maybe something else, but NullAway seems to be determining that `io.reactivex.functions.Function.apply` (the functional interface for the second lambda) has an `@NonNull` return. 

Then it's interpreting `optional.get()` as `@Nullable` due to the [OptionalEmptinessHandler](https://github.com/uber/NullAway/blob/d2d84b0114d394832e8ce904cf94aa0d9ad5639a/nullaway/src/main/java/com/uber/nullaway/handlers/OptionalEmptinessHandler.java#L63).

This causes this case to trigger when evaluating the lambda's [return](https://github.com/uber/NullAway/blob/master/nullaway/src/main/java/com/uber/nullaway/NullAway.java#L641).

**Comment by lazaroclapp** [comment]
I'd say it needs only a `handler.onPrepareErrorMessage(...)` right after this line and we should be good: https://github.com/uber/NullAway/blob/master/nullaway/src/main/java/com/uber/nullaway/NullAway.java#L642 :)

**Comment by lazaroclapp** [comment]
Note, however, that there might be a deeper issue. What happens with the `OptionalEmptinessHandler` if it sees code like this:

```
Optional foo() {
   return Optional.empty();
}
```

I believe the handler will forbid it. This might be what we want, but that might require a more appropriate error message. If so, we also need to handle the errors that should be triggered on e.g. field initialization:

```
private Optional bar = Optional.empty();
```

And so forth.

**Comment by msridhar** [comment]
Sorry what does this one have to do with method references?  Is the issue title out of date?

**Comment by lazaroclapp** [comment]
> Sorry what does this one have to do with method references?  Is the issue title out of date?

The title of the issue and the body of the issue are both true, but different, bugs :)

**Comment by shas19** [comment]
> Sorry what does this one have to do with method references? Is the issue title out of date?

Yeah 😅.. These are two separate issues.

**Comment by shas19** [comment]
> I'd say it needs only a `handler.onPrepareErrorMessage(...)` right after this line and we should be good: https://github.com/uber/NullAway/blob/master/nullaway/src/main/java/com/uber/nullaway/NullAway.java#L642 :)

Thanks! this works. #309 might require more changes though

**Comment by msridhar** [comment]
Regarding @lazaroclapp's comment [above](https://github.com/uber/NullAway/issues/307#issuecomment-486463533), I think one issue is that we have no way to annotate an `Optional` type as present or not present.  The corresponding CF checker [has these annotations](https://checkerframework.org/manual/#optional-annotations).  I think cases where the annotations are required are rare enough that adding support should be low priority, but we may want to document the limitations of our current checking and where users may have to just suppress warnings for now.
